tuned_lens/nn/unembed.py

```diff
@@ -21,76 +20,34 @@ class InversionOutput:
     kl: th.Tensor
     loss: th.Tensor
     nfev: int
-    hessian: Optional[th.Tensor]
 
 
 class Unembed(th.nn.Module):
     """Module that maps transformer hidden states to logits (and vice versa)."""
 
-    transformer_layers: th.nn.ModuleList
-    layer_norm: th.nn.LayerNorm
+    final_norm: model_surgery.Norm
     unembedding: th.nn.Linear
 
-    _U_pinv: th.Tensor
-
     def __init__(
         self,
         model: PreTrainedModel,
-        extra_layers: int = 0,
     ):
         """Initialize unmebed.
 
         Args:
             model: A HuggingFace model from which to extract the unembedding matrix.
-            extra_layers: To leave at the end of the transformer.
         """
         super().__init__()
-        self.transformer_layers = th.nn.ModuleList()
-        # Initializing from scratch without a model
-
-        raw_ln = get_final_norm(model)
-        assert raw_ln is not None
+        final_norm = model_surgery.get_final_norm(model)
 
-        raw_unembed = model.get_output_embeddings()
-        if not isinstance(raw_unembed, th.nn.Linear):
+        unembeding_matrix = model.get_output_embeddings()
+        if not isinstance(unembeding_matrix, th.nn.Linear):
             # With nn.Linear we know that the unembedding matrix is .weight;
             # we don't want to guess incorrectly for other module classes.
             raise ValueError("Currently we only support nn.Linear unembeddings.")
 
-        U = raw_unembed.weight.data.float()
-        assert isinstance(U, th.Tensor)
-
-        vocab_size, d_model = U.shape
-        self.layer_norm = th.nn.LayerNorm(
-            d_model, elementwise_affine=False, eps=getattr(raw_ln, "eps", 1e-5)
-        )
-        self.unembedding = th.nn.Linear(d_model, vocab_size, device=U.device)
-
-        # Roll the LN bias into our unembedding Linear
-        gamma, beta = raw_ln.weight.data.float(), raw_ln.bias.data
-        if isinstance(beta, th.Tensor):
-            bias = beta.float() @ U.T
-
-            # GPT-J has a bias in the unembedding layer, so we need to add it
-            if raw_unembed.bias is not None:
-                bias += raw_unembed.bias.data.float()
-
-            self.unembedding.bias.data = bias.to(beta.dtype)
-
-        # Roll the LN diagonal scaling factor into our unembedding matrix
-        if isinstance(gamma, th.Tensor):
-            U = U * gamma
-
-        self.unembedding.weight.data = U.to(raw_unembed.weight.dtype)
-
-        # We precompute the pseudo-inverse of the unembedding matrix
-        self.register_buffer("_U_pinv", U.pinverse())
```

Comment:
Could probably keep this

Reply:
Don't think its worth keeping. Seems complicated and it was never actually being tested. If the speed of this operation ends up being a problem for future work we can consider reimplementing it.